test/Dialect/LLHD/Transforms/mem2reg.mlir

```diff
@@ -493,5 +493,61 @@ hw.module @ConditionalDrives(in %u: i42, in %v: i42, in %q: i1, in %r: i1) {
   }
 }
 
+// Delayed and blocking drive interaction.
+// CHECK-LABEL: @DelayedDrives
+hw.module @DelayedDrives(in %u: i42, in %v: i42, in %bool: i1) {
+  %0 = llhd.constant_time <0ns, 0d, 1e>
+  %1 = llhd.constant_time <0ns, 1d, 0e>
+  %a = llhd.sig %u : i42
+  // Delayed drives after blocking drives persist.
+  // CHECK: llhd.process
+  llhd.process {
+    // CHECK-NEXT: [[T:%.+]] = llhd.constant_time <0ns, 0d, 1e>
+    // CHECK-NEXT: llhd.drv %a, %u after [[T]]
+    // CHECK-NEXT: [[T:%.+]] = llhd.constant_time <0ns, 1d, 0e>
+    // CHECK-NEXT: llhd.drv %a, %v after [[T]]
+    llhd.drv %a, %u after %0 : !hw.inout<i42>
+    llhd.drv %a, %v after %1 : !hw.inout<i42>
+    // CHECK-NEXT: llhd.halt
+    llhd.halt
+  }
+  // Later blocking drives erase earlier delayed drives.
+  // CHECK: llhd.process
+  llhd.process {
+    // CHECK-NOT: llhd.drv %a, %u
+    // CHECK-NEXT: [[T:%.+]] = llhd.constant_time <0ns, 0d, 1e>
+    // CHECK-NEXT: llhd.drv %a, %v after [[T]]
+    llhd.drv %a, %u after %1 : !hw.inout<i42>
+    llhd.drv %a, %v after %0 : !hw.inout<i42>
+    // CHECK-NEXT: llhd.halt
+    llhd.halt
+  }
+  // CHECK: llhd.process
+  llhd.process {
+    // CHECK-NOT: llhd.drv
+    llhd.drv %a, %u after %1 : !hw.inout<i42>
+    // CHECK-NEXT: hw.constant 0 : i42
+    // CHECK-NEXT: hw.constant false
+    // CHECK-NEXT: cf.cond_br %bool, ^bb1, ^bb2({{%c0_i42.*}}, {{%false.*}}, %u, {{%true.*}} : i42, i1, i42, i1)
+    cf.cond_br %bool, ^bb1, ^bb2
+  ^bb1:
+    // CHECK-NEXT: ^bb1:
+    // CHECK-NOT: llhd.drv
+    llhd.drv %a, %v after %0 : !hw.inout<i42>
+    // CHECK-NEXT: hw.constant 0 : i42
+    // CHECK-NEXT: hw.constant false
+    // CHECK-NEXT: cf.br ^bb2(%v, {{%true.*}}, {{%c0_i42.*}}, {{%false.*}} : i42, i1, i42, i1)
+    cf.br ^bb2
+  ^bb2:
+    // CHECK-NEXT: ^bb2([[ABLK:%.+]]: i42, [[ABLKCOND:%.+]]: i1, [[ADEL:%.+]]: i42, [[ADELCOND:%.+]]: i1):
+    // CHECK-NEXT: [[T:%.+]] = llhd.constant_time <0ns, 0d, 1e>
+    // CHECK-NEXT: llhd.drv %a, [[ABLK]] after [[T]] if [[ABLKCOND]]
+    // CHECK-NEXT: [[T:%.+]] = llhd.constant_time <0ns, 1d, 0e>
+    // CHECK-NEXT: llhd.drv %a, [[ADEL]] after [[T]] if [[ADELCOND]]
+    // CHECK-NEXT: llhd.halt
+    llhd.halt
+  }
+}
```

Comment:
If there are two delayed drives to the same signal some conflict resolution has to happen AFAIK, but in mem2reg we probably just want to leave it as is and not merge the drives. Can you add a test to check what mem2reg does in such a situation?

Reply:
If I remember correctly, this only happens on wires, where you cannot have procedural drives assign to them. If you have two drives to the same variable, the latter one should win and override the earlier one. Let me go check on the details

Reply:
10.4.2 in IEEE the 1800-2007 standard indicates that driving the same variable from different processes will see one of the assignments win but no conflict resolution occurs:

<img width="814" alt="Screen Shot 2025-02-27 at 08 53 23" src="https://github.com/user-attachments/assets/b3f0d91a-1aff-4b74-807e-a7f8c303b11c" />

In the same section, there's an example showing how delayed drives with a shorter delay override already scheduled delayed drives with a longer delay:

<img width="809" alt="Screen Shot 2025-02-27 at 08 54 41" src="https://github.com/user-attachments/assets/a278124d-e269-420c-b9e9-c6a34e641907" />